tailnet/controllers.go

```diff
@@ -871,43 +871,51 @@ type TunnelAllWorkspaceUpdatesController struct {
 	updater *tunnelUpdater
 }
 
-type workspace struct {
-	id     uuid.UUID
-	name   string
-	agents map[uuid.UUID]agent
+type Workspace struct {
+	ID     uuid.UUID
+	Name   string
+	Status proto.Workspace_Status
+
+	ownerUsername string
+	agents        map[uuid.UUID]*Agent
 }
 
-// addAllDNSNames adds names for all of its agents to the given map of names
-func (w workspace) addAllDNSNames(names map[dnsname.FQDN][]netip.Addr, owner string) error {
-	for _, a := range w.agents {
+// updateDNSNames updates the DNS names for all agents in the workspace.
+func (w *Workspace) updateDNSNames() error {
```

Comment:
Including all the FQDNs as part of an outgoing workspace update was kinda tricky. 

This solution just has this function update the hosts on each agent, and then store all the agents and workspaces as pointers. This lets us build our workspace update as we apply the diff to the state, and then before sending, we fill in all the correct FQDNs, which applies to both the state, and the outgoing update.